dgisim/tests/test_dices.py

```diff
@@ -167,101 +168,438 @@ def test_just_satisfy_typical_all_omni(self):
+            for b in range(a + 1, len(elements_and_omni)):
+                # need = supply, need < supply, need > supply
+                for any_need_number in [(b - a + 1), (b - a), (b - a + 2)]:
+                    with self.subTest(a=a, b=b, any_need_number=any_need_number):
+                        input_actual_dices: ActualDices = ActualDices(
+                            dices={el: 1 for el in elements_and_omni[a:b + 1]}
+                        )
+                        input_abstract_dices: AbstractDices = AbstractDices(
+                            dices={Element.ANY: any_need_number}
+                        )
+                        input_precedence: list[set[Element]] = [
+                            {el} for el in Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY
+                        ]
+
+                        actual: Optional[ActualDices] = input_actual_dices.smart_selection(
+                            requirement=input_abstract_dices, game_state=None,
+                            local_precedence_arg=input_precedence
+                        )
+
+                        if any_need_number == (b - a + 1):  # =
+                            expected: Optional[ActualDices] = input_actual_dices
+                        elif any_need_number == (b - a):  # need < supply
+                            expected = ActualDices(
+                                dices={el: 1 for el in elements_and_omni[a + 1:b + 1]})
+                        elif any_need_number == (b - a + 2):
+                            expected = None
+                        self.assertEqual(actual, expected)
+
+    def test_smart_selection_concrete_element(
+            self, element: Element = Element.GEO):
+        """test fill element with itself
+        needed=1..10
+        supply=1..10"""
+
+        for needed in range(1, 11):
+            for supply in range(1, 11):
+                with self.subTest(needed=needed, supply=supply):
+                    input_actual_dices: ActualDices = ActualDices(dices={element: supply})
+                    input_abstract_dices: AbstractDices = AbstractDices(
+                        dices={element: needed}
+                    )
+                    input_precedence: list[set[Element]] = []
+
+                    actual = input_actual_dices.smart_selection(
+                        requirement=input_abstract_dices,
+                        game_state=None,
+                        local_precedence_arg=input_precedence
+                    )
+
+                    if supply >= needed:
+                        expected = ActualDices(dices={element: needed})
+                    else:
+                        expected = None
+
+                    self.assertEqual(actual, expected)
+
+    def test_smart_selection_precedence(self):
+        """basic test on elements local precedence"""
+        element_geo = Element.GEO
+        element_pyro = Element.PYRO
+        for any_needed_number in range(5, 16, 5):
```

Comment:
I don't quite get why there are two `5`s here

Reply:
to check 5, 10, 15 (3rd argument in slice is step). Made it more clear

Reply:
My bad, didn't look carefully enough, thought it's a list